From 00aa72e5523df52df96322994bfb76474409844e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:06:01 -0400
Subject: [usb] usbfs: Kill urbs on interface before doing a set_interface

Message-id: <1410887463-20674-290-git-send-email-dzickus@redhat.com>
Patchwork-id: 95230
O-Subject: [RHEL7 PATCH 289/591] usbfs: Kill urbs on interface before doing a set_interface
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 5ec9c1771ce83a1e2b7ec96ed9f29a9f1b25e71e
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Wed Oct 9 17:19:27 2013 +0200

    usbfs: Kill urbs on interface before doing a set_interface

    The usb_set_interface documentation says:

     * Also, drivers must not change altsettings while urbs are scheduled for
     * endpoints in that interface; all such urbs must first be completed
     * (perhaps forced by unlinking).

    For in kernel drivers we trust the drivers to get this right, but we
    cannot trust userspace to get this right, so enforce it by killing any
    urbs still pending on the interface.

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/core/devio.c b/drivers/usb/core/devio.c
index f3ba2e0..2a95e4e 100644
--- a/drivers/usb/core/devio.c
+++ b/drivers/usb/core/devio.c
@@ -1143,6 +1143,9 @@ static int proc_setintf(struct dev_state *ps, void __user *arg)
   return -EFAULT;
  if ((ret = checkintf(ps, setintf.interface)))
   return ret;
+
+ destroy_async_on_interface(ps, setintf.interface);
+
  return usb_set_interface(ps->dev, setintf.interface,
    setintf.altsetting);
 }
-- 
1.7.1